File: scripts/pdf_fixture.py
from __future__ import annotations

from pathlib import Path


def _escape_pdf_text(value: str) -> str:
    return value.replace("\\", "\\\\").replace("(", "\\(").replace(")", "\\)")


def build_single_page_text_pdf(text: str) -> bytes:
    lines = [line.strip() for line in text.splitlines() if line.strip()]
    if not lines:
        raise ValueError("PDF fixture text must not be empty")
    commands = ["BT", "/F1 11 Tf", "72 760 Td"]
    for index, line in enumerate(lines):
        if index:
            commands.append("0 -18 Td")
        commands.append(f"({_escape_pdf_text(line)}) Tj")
    commands.append("ET")
    stream = ("\n".join(commands) + "\n").encode("ascii")

    objects = [
        b"<< /Type /Catalog /Pages 2 0 R >>",
        b"<< /Type /Pages /Kids [3 0 R] /Count 1 >>",
        (
            b"<< /Type /Page /Parent 2 0 R /MediaBox [0 0 612 792] "
            b"/Resources << /Font << /F1 4 0 R >> >> /Contents 5 0 R >>"
        ),
        b"<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica >>",
        b"<< /Length "
        + str(len(stream)).encode("ascii")
        + b" >>\nstream\n"
        + stream
        + b"endstream",
    ]

    output = bytearray(b"%PDF-1.7\n%\xe2\xe3\xcf\xd3\n")
    offsets = [0]
    for number, body in enumerate(objects, start=1):
        offsets.append(len(output))
        output.extend(f"{number} 0 obj\n".encode("ascii"))
        output.extend(body)
        output.extend(b"\nendobj\n")
    xref_offset = len(output)
    output.extend(f"xref\n0 {len(objects) + 1}\n".encode("ascii"))
    output.extend(b"0000000000 65535 f \n")
    for offset in offsets[1:]:
        output.extend(f"{offset:010d} 00000 n \n".encode("ascii"))
    output.extend(
        (
            f"trailer\n<< /Size {len(objects) + 1} /Root 1 0 R >>\n"
            f"startxref\n{xref_offset}\n%%EOF\n"
        ).encode("ascii")
    )
    return bytes(output)


def build_fixture(path: Path) -> bytes:
    return build_single_page_text_pdf(path.read_text(encoding="utf-8"))

File: tests/e2e/pdf-fixture.ts
import { readFileSync } from "node:fs";

function escapePdfText(value: string): string {
  return value
    .replaceAll("\\", "\\\\")
    .replaceAll("(", "\\(")
    .replaceAll(")", "\\)");
}

function append(
  output: Buffer<ArrayBufferLike>,
  value: Buffer<ArrayBufferLike> | string,
): Buffer<ArrayBufferLike> {
  return Buffer.concat([
    output,
    typeof value === "string" ? Buffer.from(value, "ascii") : value,
  ]);
}

export function buildSinglePageTextPdf(text: string): Buffer<ArrayBufferLike> {
  const lines = text
    .split(/\r?\n/u)
    .map((line) => line.trim())
    .filter(Boolean);
  if (lines.length === 0) {
    throw new Error("PDF fixture text must not be empty");
  }

  const commands = ["BT", "/F1 11 Tf", "72 760 Td"];
  lines.forEach((line, index) => {
    if (index > 0) {
      commands.push("0 -18 Td");
    }
    commands.push(`(${escapePdfText(line)}) Tj`);
  });
  commands.push("ET");
  const stream = Buffer.from(`${commands.join("\n")}\n`, "ascii");
  const objects = [
    Buffer.from("<< /Type /Catalog /Pages 2 0 R >>", "ascii"),
    Buffer.from("<< /Type /Pages /Kids [3 0 R] /Count 1 >>", "ascii"),
    Buffer.from(
      "<< /Type /Page /Parent 2 0 R /MediaBox [0 0 612 792] " +
        "/Resources << /Font << /F1 4 0 R >> >> /Contents 5 0 R >>",
      "ascii",
    ),
    Buffer.from(
      "<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica >>",
      "ascii",
    ),
    Buffer.concat([
      Buffer.from(`<< /Length ${stream.length} >>\nstream\n`, "ascii"),
      stream,
      Buffer.from("endstream", "ascii"),
    ]),
  ];

  let output: Buffer<ArrayBufferLike> = Buffer.from("%PDF-1.7\n%", "ascii");
  output = append(output, Buffer.from([0xe2, 0xe3, 0xcf, 0xd3]));
  output = append(output, "\n");
  const offsets = [0];
  objects.forEach((body, index) => {
    offsets.push(output.length);
    output = append(output, `${index + 1} 0 obj\n`);
    output = append(output, body);
    output = append(output, "\nendobj\n");
  });
  const xrefOffset = output.length;
  output = append(output, `xref\n0 ${objects.length + 1}\n`);
  output = append(output, "0000000000 65535 f \n");
  offsets.slice(1).forEach((offset) => {
    output = append(
      output,
      `${offset.toString().padStart(10, "0")} 00000 n \n`,
    );
  });
  output = append(
    output,
    `trailer\n<< /Size ${objects.length + 1} /Root 1 0 R >>\n` +
      `startxref\n${xrefOffset}\n%%EOF\n`,
  );
  return output;
}

export function canonicalInvoicePdf(): Buffer<ArrayBufferLike> {
  return buildSinglePageTextPdf(
    readFileSync("tests/fixtures/canonical_invoice.txt", "utf8"),
  );
}
